### ISSUE TITLE
Not an Issue, just a tip and how I do this with a shelly 3EM

### ISSUE THREAD

**Issue by bemilamedia** [That-Dude/Emulated-Eastrom-SDM-Power-Meter#1]
Hi that-Dude,
thanks for this project. I have installed it on my ESP an I'm waiting on my TTTL-RS485 converter. 
And I have to finish the installation of my solarsystem at first for a real world test. :-) (Hope this works with my Growatt MIC3000TL-X. But I think it does)

I have played a little bit arround and I found the following in line 136

`// Subscribe
      client.subscribe("shellies/shellyem-house/emeter/0/power");`

I have changed it to 
`// Subscribe
      client.subscribe(mqtt_topic);`
So it is directly linked to the value of mqtt-topic (in private.h). When I first compiled it, I wondered why it didn't work. Only then did I see that the "link" to the shelly was manually entered here again. It's not a problem, but it reduces the error of having to change this manually in line 136 :-)

In my case I use a shelly 3EM to measure all 3 phases of my home. The big and known problem of the 3EM is, he doesn't calculate the balance over all phases. I do this calculation in Home Assistant with:

`bilanz_bezug_solarproduktion:
        friendly_name: "aktuelle_Bilanz_Bezug_Solarproduktion"
        unit_of_measurement: 'W'
        value_template: >-
            {{ states("sensor.shellyem3_c45bbe785f3e_channel_a_power") |float(0) |round(0)
             + states("sensor.shellyem3_c45bbe785f3e_channel_b_power") |float(0) |round(0) 
             + states("sensor.shellyem3_c45bbe785f3e_channel_c_power") |float(0) |round(0) }}
        availability_template: "{{
            [ states('sensor.shellyem3_c45bbe785f3e_channel_a_power'),
              states('sensor.shellyem3_c45bbe785f3e_channel_b_power'),
              states('sensor.shellyem3_c45bbe785f3e_channel_c_power')
            ] | map('is_number') | min
          }}"`

Next I create a Node-Red node with poll-state of the "sensor.bilanz_bezug_solarproduktion" every 15 seconds and send it as mqtt-topic "shellies/Saldierung_Zaehler/Power". In the private.h I insert
`const char* mqtt_topic = "shellies/Saldierung_Zaehler/Power";  // MQTT topic to get power reading from
`
This works fine at the moment. I have positive values in my ESP when I get power from the net. And negative values when I produce more power then I use.

**Comment by That-Dude** [comment]
Nice catch @bemilamedia  - I've fixed it.

It occurs to me that it would be quite easy to grab all three Shelly 3EM topics and do that balance calculation on the ESP8266 taking Home Assistant out of the loop. I love HA but it's another moving part waiting to go wrong in this scenario.

**Comment by bemilamedia** [comment]
Yes, I have this idea too. To make the calculation directly with the ESP. This was the easiest way at the moment without programming anything. Because I already have this calculated value in HA for visualization. So it is easy to transfer this value into MQTT. But yes, there can be errors because there are 2 more software parts. HA and NodeRed. 

I will try to implement this once in ESP. Because then I can also output the balancing from ESP to MQTT and visualize it in HA. :-)

**Comment by Feierprinz** [comment]
Hi guys,

I´m very interested to create these for Shelly3EM and in future my Growatt SPH-Inverter.
The Shelly3EM has an API to receive the values:
http://xxx.xxx.xxx.xxx/emeter/0
http://xxx.xxx.xxx.xxx/emeter/1
http://xxx.xxx.xxx.xxx/emeter/2

on Each URL we receive
power (420.54) in Watt
pf (Powerfactor)
current (2.06) in Ampere
and the voltage (226.35) in Volt

Is it not simple to recalculate the received Power and transmit it to the RS485? Have you an idea how can I emulate a SDM630 with your great Repo?

**Comment by bemilamedia** [comment]
Hi zusammen,
ich habe den Code etwas umgeschrieben und ich lese alle 3 Phasen des 3EM per MQTT ein. Diese saldiere ich dann im ESP8266 und schicke dann diesen Wert raus.

Praktisch habe ich den Code noch nicht testen können, da sich das Projekt etwas hinzieht. (Knieverletzung und ich kann die Solarodule derzeit nicht montieren)

Ich schaue mal wie ich einen Fork erstelle und packe dann meine Anpassung da rein.

**Comment by bemilamedia** [comment]
Hi together,
I have rewritten the code a bit and I am reading in all 3 phases of the 3EM via MQTT. I then balance these in the ESP8266 and then send this value out.

Practically I have not yet been able to test the code, because the project drags on a bit. (Knee injury and I can't mount the solar modules at the moment).

I'll see how I create a fork and then put my customization in there.

**Comment by bemilamedia** [comment]
UPDATE:
I have created and updated "my" fork with all the changes to work with the Shelly 3EM, picking up the MQTT topics for each phase, calculate the balanced power and give this value out.

https://github.com/bemilamedia/Emulated-Eastrom-SDM-Power-Meter-with-Shelly-3EM

(Hope it works - untested)
